### ISSUE TITLE
flycheck with verilator: non-zero exit code, but no errors

### ISSUE THREAD

**Issue by Dicridon** [flycheck/flycheck#1250]
![screenshot from 2017-05-08 10-46-31](https://cloud.githubusercontent.com/assets/24477137/25788524/5feb9198-33dc-11e7-80c7-ea1b54080a47.png)
Hi, I'm a student trying to use emacs. Flycheck helped me checking syntax. Recently, I am using verilog, and I want flycheck to check verilog code, so I follow your instructtion and installed verilator.
But flycheck report: non-zero exit code, but no errors. I followed the tips and installed latest verilator, but the same problem still exists. Actually I am not sure if this is a bug, so I just follow this guide and report this issue to you.
* What you did, and what you expected to happen instead
I installed verilator for verilog syntax checking, but flycheck reported "non-zero exit code, but no errors"
So I looked through troubleshooting guide, and installed latest verilator, but the same problem still exists.
* Whether and how you were able to [reproduce the issue in emacs -Q][emacsQ]
I am new to emacs and have no idea how to use emacs -Q
* Your Flycheck setup from `M-x flycheck-verify-setup`
I did so, but the same error still appears as shown in my picture
* Your operating system
uname output: Linux, linux.suse 4.4.62-18.6-default
* Your Emacs version from `M-x emacs-version`
GNU Emacs 24.3.1
* Your Flycheck version from `M-x flycheck-version`
31alpha

Really wish you could help with it, thank you~

**Comment by fmdkdd** [comment]
Thanks for the report!  `flycheck-verify-setup` will only give you information about your checkers configuration; it doesn't try to fix anything.  Here you can see that everything is green, so that's a good start.

I would not be surprised if the `verilog-verilator` checker we use in Flycheck was out of date.  Can you report the output of `M-x flycheck-compile` when you are in a Verilog buffer?

**Comment by CeleritasCelery** [comment]
I have noticed this same behavior.

**Comment by fmdkdd** [comment]
@CeleritasCelery Do you have the same error "Can't exec program: /usr/local/bin/verilator"?  Because that looks like a path issue. 

Please report your the output of `M-x flycheck-compile` so we have more information.

**Comment by djg** [comment]
I have the same issue.  When I save the file and flycheck runs, I get the following error message:

```
Suspicious state from syntax checker verilog-verilator: Flycheck checker verilog-verilator returned non-zero exit code 126, but its output contained no errors: emacs: /home/djg/verilator-2018-03-04/bin/verilator: Exec format error

Try installing a more recent version of verilog-verilator, and please open a bug report if the issue persists in the latest release.  Thanks!
```

If I run the flycheck command with `M-x flycheck-compile`, it works:

```
verilator --lint-only -Wall /home/djg/Work/xpu/rtl/common/common.v
%Error: /home/djg/Work/xpu/rtl/common/common.v:7: syntax error, unexpected reg, expecting ';'
%Error: /home/djg/Work/xpu/rtl/common/common.v:9: syntax error, unexpected assign
%Error: /home/djg/Work/xpu/rtl/common/common.v:20: syntax error, unexpected input, expecting ')' or ','
%Error: Exiting due to 3 error(s)
%Error: Command Failed /home/djg/verilator-2018-03-04/bin/verilator_bin --lint-only -Wall /home/djg/Work/xpu/rtl/common/common.v

Compilation exited abnormally with code 10 at Wed Apr 18 17:42:32
```

Could the commands be launched differently between the two invocations? I notice that verilator is a perl script that has a non-standard shell script first line. (ie. it's missing the `shebang`):

```sh
: # -*-Mode: perl;-*- use perl, wherever it is
```

**Comment by fmdkdd** [comment]
@djg Interesting.  There is a difference between `flycheck-compile` and `flycheck-buffer`: the former calls `verilator` in a shell, while the latter uses `(flycheck-find-checker-executable 'verilog-verilator)` to find the executable.

So, depending on how your PATH is set in your shell and in Emacs, that could explain the different calls.  One way to check that is to evaluate `M-: (flycheck-find-checker-executable 'verilog-verilator)` in Emacs, and `which verilator` in your shell.

**Comment by djg** [comment]
Changing the verilator command from `verilator` to `verilator_bin` in `flycheck.el` works.

**Comment by fmdkdd** [comment]
> Changing the verilator command from verilator to verilator_bin in flycheck.el works.

Instead of modifying `flycheck.el` directly, you can set `flycheck-verilog-verilator-executable`.

**Comment by cpitclaudel** [comment]
This was a path issue, most likely;  in the meantime, I've created a PR to update verilator patterns at https://github.com/flycheck/flycheck/pull/1691
